torch/_inductor/codegen/cpp.py

```diff
@@ -3372,13 +3377,107 @@ def get_kernel_group(self):
         else:
             self.kernel_group = KernelGroup()
 
+    def fuse(self, node1, node2):
+        if node1.is_foreach() or node2.is_foreach():
+            return ForeachKernelSchedulerNode.fuse(node1, node2)
+        else:
+            if self._can_fuse_nodes_with_compatible_ranges(node1, node2):
+                assert isinstance(node1, (SchedulerNode, FusedSchedulerNode))
+                assert isinstance(node2, (SchedulerNode, FusedSchedulerNode))
+
+                _, (vars1, reduce1) = node1.group
+                _, (vars2, reduce2) = node2.group
+                assert reduce1 == () and reduce2 == (), (reduce1, reduce2)
+
+                def get_indexing_ranges_exprs(node):
+                    if isinstance(node, FusedSchedulerNode):
+                        assert len(node.snodes) > 0
+                        # use the last scheduler node from the list as it has the most
+                        # relevant indexing expressions
+                        return get_indexing_ranges_exprs(node.snodes[-1])
+                    else:
+                        assert isinstance(node, SchedulerNode)
+                        comp_buffer = node.node
+                        assert isinstance(comp_buffer, ir.ComputedBuffer)
+                        _, body, _ = comp_buffer.get_default_sizes_body()
+                        return body.var_ranges, [*body.indexing_exprs.values()]
+
+                node_to_recomp = node1 if len(vars1) < len(vars2) else node2
+                assert isinstance(node_to_recomp, SchedulerNode)
+
+                ref_node = node2 if len(vars1) < len(vars2) else node1
+
+                extra_indexing_constraints = get_indexing_ranges_exprs(ref_node)
+
+                node_to_recomp.recompute_size_and_body(
+                    extra_indexing_constraints=extra_indexing_constraints
+                )
+
+                if len(vars1) < len(vars2):
+                    node1 = node_to_recomp
+                else:
+                    node2 = node_to_recomp
+
+                _, (vars1, _) = node1.group
+                _, (vars2, _) = node2.group
+                assert vars1 == vars2, (vars1, vars2)
+
+            return FusedSchedulerNode.fuse(node1, node2)
+
+    def _can_fuse_nodes_with_compatible_ranges(self, node1, node2):
+        # Here we try to fuse SchedulerNode/FusedSchedulerNode with compatible ranges
+        # e.g. (s0, s1, s2) and (s0 * s1 * s2)
+        _, (vars1, reduce1) = node1.group
+        _, (vars2, reduce2) = node2.group
+
+        c1 = reduce1 == () and reduce2 == ()
+        c2 = math.prod(vars1) == math.prod(vars2)
+        c3 = len(vars1) == 1 or len(vars2) == 1
+        if not (c1 and c2 and c3):
+            return False
+
+        node_to_recomp = node1 if len(vars1) < len(vars2) else node2
+        ref_node = node2 if len(vars1) < len(vars2) else node1
+
+        # We can recompute sizes and body for nodes other than SchedulerNode
```

Comment:
Say we have:
```
fuse(FusedSchedulerNode([buf0, buf1]), FusedSchedulerNode([buf2, buf3]))
```
I think this should be equivalent to 
```
fuse(buf0, fuse(buf1, FusedSchedulerNode([buf2, buf3]))
```
or to
```
fuse(fuse(FusedSchedulerNode([buf0, buf1]), buf2), buf3)
```

So I don't see why we can't handle `FusedSchedulerNode` here.

Reply:
yes, it works like that in theory, but in practice I have to see how to handle when node_to_recomp is FusedSchedulerNode. As said previously, this is out of the scope of this PR and can be a TODO.